src/program.rs

```diff
@@ -90,7 +86,24 @@ impl<App: Application> fmt::Display for ProgramNode<App> {
             Term::Fail(..) => f.write_str("fail")?,
             Term::Jet(j) => write!(f, "[jet]{}", j)?,
         }
-        write!(f, ": {} → {}", self.source_ty, self.target_ty,)
+        write!(f, ": {} → {}", self.typed.source_ty, self.typed.target_ty,)
+    }
+}
+
+impl<App: Application> ProgramNode<App> {
+    /// Return the underlying term of the node.
+    pub fn term(&self) -> &Term<Value, App> {
+        &self.typed.term
+    }
+
+    /// Return the source type of the node.
+    pub fn source_ty(&self) -> &Arc<FinalType> {
```

Comment:
In 224e081c056fc608d00f7563faec2089320fe028:

It feels a bit unidiomatic to return `&Arc`. Probably better to just return `&FinalType` unless we have a good reason to expose internals.

Reply:
Ah, in c010975498dee1db8d1654ea7b2fd47bf78a653b you actually clone these arcs to use them as hash keys. One alternate might be to just extract the TMRs, but what you've done here is good.

Reply:
This is the same problem as with [get_left() and get_right()](https://github.com/ElementsProject/rust-simplicity/pull/30#discussion_r921191530); both actually stem from the same original commit. I prefer your solution of using the TMR because it seems Simplicity-idiomatic. `Tmr` implements `Copy`, so this should be fast.